app/components/elements/form/form/index.js

```diff
@@ -0,0 +1,106 @@
+import {cloneElement} from 'preact'
+import {path, reduce, equal} from 'wasmuth'
+import {dispatch} from '/store'
+import {
+  formSet as formSetBuilder,
+  formRemove as formRemoveBuilder
+} from '../actions.js'
+import compose from '/util/compose'
+import mapStateToProps from '/util/mapStateToProps'
+import BaseForm from './base'
+
+export default mapStateToProps(
+  (state, {name}) => path(['forms', name], state)
+)(compose({
+  componentWillMount () {
+    this.formSet = formSetBuilder(this.props.name)
+    this.formRemove = formRemoveBuilder(this.props.name)
+  },
+  componentWillUnmount () {
+    !this.props.keepFormDataOnUnmount && dispatch(this.formRemove())
+  },
+  componentWillUpdate ({
+    initialData,
+    validations = {},
+    onSubmit,
+    name,
+    data,
+    initialized,
+    asyncErrors,
+    submit,
+    errors: existingErrors
+  }) {
+    if (!initialized) {
+      dispatch(this.formSet(['initialized'], true))
+      dispatch(this.formSet(['data'], initialData))
+    }
+
+    if (submit) {
```

Comment:
I still don't get this pattern. onSubmit is a passed in prop? Why not just call it where I want (if I really need to submit a form, outside of the actual form DOM)?

Reply:
Yeah might want to rethink this part. But here's the thinking behind it.
If you have some button somewhere that needs to submit a form you just call `dispatch(set(['forms', formName, 'submit'], true))` and the form will call its onSubmit prop (with data, onSuccess, and onError args, which come with some convenient options for what to do with the form state after the api request, or whatever, is done).
so an alternative would be something like
```
onClick: submitForm(formName, (data, onSuccess, onError) => {})
or
onClick: submitForm(formName, formNameSubmitFunction)
```
Is that more like what you were thinking?